API Refactor form validation handling

tractorcow opened the pull request:
Resolves #6228
Includes rebase of #3937 as a separate commit

Merge with:
 - [ ] https://github.com/silverstripe/silverstripe-siteconfig/pull/43
 - [ ] https://github.com/silverstripe/silverstripe-asset-admin/pull/322
 - [ ] https://github.com/silverstripe/silverstripe-cms/pull/1691

The major features of this change:

 - When a form fails validation in a react admin area, it's possible for only error messages to be sent in the response, meaning simpler and more lightweight http responses.
 - If a ValidationException is thrown in any form action (e.g. in a DataObject::save) then these will now be safely caught by the Form and will be setup as nice error messages.
 - ValidationResult is now serialiseable 
 - Form messages can now be individually cast as either html / plain text
 - Form and FormField message methods moved into a `FormMessage` trait since the code was duplicated.
 - Form::setValidationResponseCallback callback now takes a parameter to the ValidationResult containing any errors.
 - Cleanup, cleanup, cleanup

sminnee commented on src/Forms/Schema/FormSchema.php:
```diff
@@ -2,10 +2,10 @@
+use SilverStripe\ORM\ValidationResult;
 
 /**
  * Represents a {@link Form} as structured data which allows a frontend library to render it.
```
This comment isn't really correct. FormSchema doesn't represent a form, because it doesn't any state. It's more like a service/factory that builds JSON representations of a Form Schema on demand. It's not related to this PR but it confused me a bit.

sminnee commented on admin/code/LeftAndMain.php:
```diff
@@ -356,37 +355,40 @@ protected function getSchemaResponse($form, $id = null) {
+		}
+		$return = ['id' => $schemaID];
 
-		if (in_array('schema', $schemaParts)) {
```
I wonder if this logic should be refactored out of `LeftAndMain` and into `FormSchema`, as it's likely to be a part of any interaction between a rich JS form and a supporting controller.

Something like this:

```php
return $this->schema->getMultipartSchema($schemaID, $schemaParts, $form, $errors);
```

Also I think that 'result' would be a better name than 'errors' for component generated by the ValidationResult, as a ValidationResult might also say "valid".

tractorcow replied:
Yes I think so. Will do that. :)

tractorcow replied:
https://github.com/silverstripe/silverstripe-framework/pull/6356/commits/368a7f0644a5f2264ceb7b2fb676b5bdeb3f0c01

tractorcow commented:
@sminnee see https://github.com/silverstripe/silverstripe-framework/pull/6356/commits/368a7f0644a5f2264ceb7b2fb676b5bdeb3f0c01 :D

chillu commented on src/Forms/Schema/FormSchema.php:
```diff
@@ -14,6 +16,47 @@
+	 * @param ValidationResult $result Required for 'error' response
+	 * @return array
+	 */
+	public function getMultipartSchema($request, $schemaID, Form $form = null, ValidationResult $result = null) {
```
I think this particular method shouldn't concern itself with *how* the schema parts are requested - if we switch this to GraphQL sometime, it likely would be contained in the GraphQL query itself, rather than `$request`. Maybe we could accept an array of parts here instead, and have another helper method to extract that from a `HTTPRequest`?

sminnee commented on docs/en/04_Changelogs/4.0.0.md:
```diff
@@ -1095,6 +1095,55 @@ Some methods on `Requirements` have had their method signatures changed:
+* `messageForForm`. removed
+* `getSessionValidationResult` / `setSessionValidationResult` used to get / set session errors
+* `getSessionData` / `setSessionData` used to get / set field values cached in the session
+* `getAjaxErrorResponse` and `getRedirectReferer` created to simplify `getValidationErrorResponse`
```
We could also make them `private`.

tractorcow replied:
Inline methods increases code complexity (of the combined block) and reduces readability. These methods are in fact not temporary; All that will change is that they will be moved to FormResponseHandler, not removed.

sminnee commented on docs/en/04_Changelogs/4.0.0.md:
```diff
@@ -1095,6 +1095,55 @@ Some methods on `Requirements` have had their method signatures changed:
+* `codeList` renamed to `getMessageCodes`
+* `message` renamed to `joinedMessage`
+* `joinedOverallMessage` added.
+* `starredList` renamed to `starredMessage`
```
Agreed. It's a hangover from the kitchen-sink mentality that SilverStripe is moving away from.

sminnee commented on docs/en/04_Changelogs/4.0.0.md:
```diff
@@ -1095,6 +1095,55 @@ Some methods on `Requirements` have had their method signatures changed:
+
+`Form` methods have been changed:
+
+* `validate` method now returns a `ValidationResult` instance. This is no longer automatically
```
Perhaps we want to:

 * Create a 2nd method with a new name. `getValidatonResult()`?
 * Rewrite `validate()` as a wrapper around this method that returns a boolean and throws a deprecation?

sminnee commented:
> Nevertheless, I'm a bit worried about the amount of API changes - given how prevalent form validation is in your average project, this could easily double the upgrade effort on some projects. We can use silverstripe-upgrader for some of this, but there's subtleties like the changed return value in validate() which will need a dev to look at things.

Can we get more of a sense of what the common use-cases are and how they need to change? I've made a specific comment about the `validate()` method above – it may be better to deprecate this method and introduce a new one rather than change its return type.

In fact, even removing the method and creating a new one would be better since then you'd get a fatal error.

Code of the form `if(!$this->validate())` would silently break, which could be dangerous.

tractorcow commented:
Updated based on feedback:

 - Form::validate() removed, now use Form::validationResponse() to generate a new `ValidationResponse` object.
 - Removed extraneous methods on ValidationResult
